File: src/TypeScript/src/hamt/persistent-ancestral-connection-forest.ts
/**
 * Fully branching persistent insertion-only connectivity forest with a first-connection query.
 *
 * `PersistentAncestralConnectionForest` is a persistent union-find over the fixed integer vertex
 * universe `0 .. vertexCount - 1`. Each successful `link` performs union by size *without* path
 * compression and labels the one new root-parent edge with the child `AncestralConnectionVersion`.
 * A redundant (already connected) link still publishes a distinct child history version, but shares
 * the complete connectivity index, which `sharesConnectivityRootWith` can confirm.
 *
 * The parent cells live sparsely in the CHAMP `PersistentHashMap`. An absent cell denotes a
 * singleton root, so `create` is O(1) rather than O(`vertexCount`). Union by size limits every
 * parent path to O(log n) cells, where n is the component size. `firstConnected` answers "at which
 * ancestor version did these two vertices first become connected?" by comparing the two current
 * parent paths and taking the latest union-edge version strictly below their forest lowest common
 * ancestor. It never searches the version history, so its cost is independent of history depth.
 *
 * Instances and version tokens are immutable snapshots. Deletion, confluent merging, retroactive
 * updates, and path compression are deliberately outside this module's contract. The O(log n)
 * parent-path factor is worst-case logarithmic in component size rather than the inverse-Ackermann
 * amortized bound of an ephemeral linear-history union-find.
 *
 * ## Why the CHAMP factor is unconditional here
 *
 * Throughout this module w denotes the cost of one parent-cell lookup or insertion in the backing
 * CHAMP map. That substrate narrows every hash-policy result to a signed 32-bit word, consumes five
 * hash bits per trie level, and parks keys sharing a full 32-bit hash in a collision node whose
 * bucket is scanned linearly. A collision bucket can therefore hold two distinct keys, and under a
 * truncating or colliding hash w would be bounded only *in expectation*.
 *
 * This forest removes that caveat by pinning `vertexHashPolicy` rather than accepting the workspace
 * default. The pinned function is the MurmurHash3 32-bit finalizer `fmix32` applied to the vertex's
 * two's-complement pattern. Every step of it is a bijection of the 32-bit word — `h ^= h >>> k` with
 * `k > 0` leaves the top `k` bits fixed and is undone by the same shape of expression, and
 * multiplication by an odd constant is invertible modulo 2**32 (both constants are odd) — and the
 * substrate's own `| 0` narrowing is a bijection too. Vertices are validated into
 * `0 .. vertexCount - 1` before any cell is touched and `maximumVertexCount` caps that universe at
 * 2**31 - 1, so the whole composition is *injective* over every key the map can ever see. No
 * collision bucket can hold two distinct vertices, and two distinct vertices must diverge within 32
 * significant hash bits, that is within `ceil(32 / 5) = 7` levels of 32-way nodes. One cell read or
 * write is therefore O(1) worst case, and every bound stated below is unconditional in both factors.
 * The accompanying test proves the injectivity by exhibiting an explicitly constructed inverse
 * rather than by sampling for absent collisions.
 *
 * ## Recursion
 *
 * Every walk here — version chains, parent paths, and the audit — is an explicit loop, so a history
 * hundreds of thousands of links deep costs no call stack. The CHAMP map's own descent recurses,
 * but over at most seven levels.
 */
import { createHashPolicy, type HashPolicy } from "./hash-policy.js";
import { PersistentHashMap } from "./persistent-hamt.js";

/**
 * Largest permitted vertex universe.
 *
 * The reference implementation's universe is a 32-bit integer, and this bound is also what keeps
 * the pinned vertex hash injective after the CHAMP map narrows it to a signed 32-bit word.
 */
export const maximumVertexCount: number = 0x7fff_ffff;

/** The MurmurHash3 32-bit finalizer, a bijection of the 32-bit word. */
function vertexHash(vertex: number): number {
    let bits = vertex | 0;
    bits ^= bits >>> 16;
    bits = Math.imul(bits, 0x85ebca6b);
    bits ^= bits >>> 13;
    bits = Math.imul(bits, 0xc2b2ae35);
    return (bits ^ (bits >>> 16)) | 0;
}

function vertexEquivalent(left: number, right: number): boolean { return left === right; }

/**
 * The hash policy this forest pins into its parent-cell map.
 *
 * Part of the type's complexity contract rather than a tuning detail: its injectivity over the
 * vertex universe is what makes the CHAMP factor w a worst-case rather than an expected constant.
 * Only injectivity is promised; the particular value is not part of the contract.
 */
export const vertexHashPolicy: HashPolicy<number> = createHashPolicy(vertexHash, vertexEquivalent);

let publishRootVersion: () => AncestralConnectionVersion;
let publishChildVersion: (parent: AncestralConnectionVersion) => AncestralConnectionVersion;

/**
 * Identifies one immutable version in a `PersistentAncestralConnectionForest` history tree.
 *
 * Tokens use *reference* identity: they deliberately have no `equals`, no `hashCode`, and no
 * `valueOf`, so `===`, `Object.is`, `Set`, and `Map` all compare identity. A token records only a
 * depth, a parent, and a root, so a structural comparison would make two sibling branches that
 * split at the same depth compare equal and let a witness from one branch masquerade as a witness
 * from the other. Use `isSameVersion` where the intent should be explicit in the reading.
 *
 * Tokens are published only by `PersistentAncestralConnectionForest.create` and `link`; the
 * constructor is private. The parent chain contains history identities, not mutable connectivity
 * state. Every accessor is O(1).
 */
export class AncestralConnectionVersion {
    readonly #parent: AncestralConnectionVersion | undefined;
    readonly #depth: number;
    readonly #root: AncestralConnectionVersion;

    static {
        publishRootVersion = (): AncestralConnectionVersion => new AncestralConnectionVersion(undefined);
        publishChildVersion = (parent: AncestralConnectionVersion): AncestralConnectionVersion =>
            new AncestralConnectionVersion(parent);
    }

    private constructor(parent: AncestralConnectionVersion | undefined) {
        if (parent !== undefined && parent.#depth >= Number.MAX_SAFE_INTEGER) {
            throw new RangeError("The connection-forest history depth would exceed the exact-integer range.");
        }
        this.#parent = parent;
        this.#depth = parent === undefined ? 0 : parent.#depth + 1;
        this.#root = parent === undefined ? this : parent.#root;
    }

    /** This version's parent, or `undefined` at the history root. O(1). */
    public get parent(): AncestralConnectionVersion | undefined { return this.#parent; }

    /** The number of link events from the history root. O(1). */
    public get depth(): number { return this.#depth; }

    /** The root identity of this version tree, cached rather than walked. O(1). */
    public get root(): AncestralConnectionVersion { return this.#root; }

    /** Whether two handles denote the very same version. O(1), and identical to `===`. */
    public isSameVersion(other: AncestralConnectionVersion): boolean { return this === other; }
}

/** Shape measurements returned by a successful structural audit. */
export interface AncestralConnectionForestStatistics {
    /** The fixed vertex-universe size. */
    readonly vertexCount: number;
    /** The number of connected components, recounted from the parent forest. */
    readonly componentCount: number;
    /** The number of explicitly stored parent cells; absent cells are singleton roots. */
    readonly storedCellCount: number;
    /** The longest parent path found, in edges. Bounded by `floor(log2(vertexCount))`. */
    readonly maximumParentPathLength: number;
    /** The number of link events from the history root to the audited version. */
    readonly historyDepth: number;
}

/**
 * One sparse parent record. `size` is meaningful only for a root, whose `parent` is the vertex
 * itself, and `joinedAt` is present only on a non-root edge. An absent cell canonically means
 * "singleton root of size 1".
 */
interface Cell {
    readonly parent: number;
    readonly size: number;
    readonly joinedAt: AncestralConnectionVersion | undefined;
}

/** One vertex on a parent path together with the union-edge tag leaving it. */
interface PathStep {
    readonly vertex: number;
    readonly joinedAt: AncestralConnectionVersion | undefined;
}

function rootCell(vertex: number, size: number): Cell { return { parent: vertex, size, joinedAt: undefined }; }

function childCell(parent: number, joinedAt: AncestralConnectionVersion): Cell {
    return { parent, size: 0, joinedAt };
}

/**
 * The deeper of two candidate versions, keeping `left` on an equal depth.
 *
 * Every tag reachable from one version's cells lies on that version's own ancestor chain, so the
 * candidates are totally ordered by depth and an actual tie cannot arise through the public API;
 * the rule is a determinism guarantee that keeps the comparison independent of accumulation order.
 */
function later(
    left: AncestralConnectionVersion | undefined,
    right: AncestralConnectionVersion | undefined,
): AncestralConnectionVersion | undefined {
    return right !== undefined && (left === undefined || right.depth > left.depth) ? right : left;
}

/**
 * A fully branching persistent insertion-only connectivity forest that reports the first ancestor
 * version in which two vertices became connected, without searching the version history.
 *
 * See the module documentation for the cost model; w below is the CHAMP path cost, which the pinned
 * `vertexHashPolicy` makes an unconditional O(1).
 */
export class PersistentAncestralConnectionForest {
    readonly #cells: PersistentHashMap<number, Cell>;
    readonly #vertexCount: number;
    readonly #componentCount: number;
    readonly #version: AncestralConnectionVersion;

    private constructor(
        vertexCount: number,
        componentCount: number,
        cells: PersistentHashMap<number, Cell>,
        version: AncestralConnectionVersion,
    ) {
        this.#vertexCount = vertexCount;
        this.#componentCount = componentCount;
        this.#cells = cells;
        this.#version = version;
    }

    /**
     * An edgeless root version over vertices `0 .. vertexCount - 1`. O(1).
     *
     * Every vertex starts as its own component. Because a singleton is represented by the *absence*
     * of a cell, construction costs the same for a universe of two vertices and for one of
     * `maximumVertexCount`. Throws `RangeError` when `vertexCount` is not an integer in
     * `0 .. maximumVertexCount`.
     */
    public static create(vertexCount: number): PersistentAncestralConnectionForest {
        if (!Number.isInteger(vertexCount) || vertexCount < 0) {
            throw new RangeError("Vertex count must be a non-negative integer.");
        }
        if (vertexCount > maximumVertexCount) {
            throw new RangeError(`Vertex count cannot exceed ${maximumVertexCount}.`);
        }
        return new PersistentAncestralConnectionForest(
            vertexCount,
            vertexCount,
            PersistentHashMap.empty<number, Cell>(vertexHashPolicy),
            publishRootVersion(),
        );
    }

    /** The fixed number of vertices in this forest. O(1). */
    public get vertexCount(): number { return this.#vertexCount; }

    /** The number of connected components in this version. O(1). */
    public get componentCount(): number { return this.#componentCount; }

    /** The identity token for this history version. O(1). */
    public get version(): AncestralConnectionVersion { return this.#version; }

    /**
     * The current union-find representative of a vertex. O(w log n) worst case.
     *
     * The representative is selected by deterministic union-by-size history and is an implementation
     * artifact: sibling versions need not agree on it. Throws `RangeError` when the vertex is
     * outside the fixed universe.
     */
    public find(vertex: number): number {
        this.#checkVertex(vertex);
        return this.#findRoot(vertex);
    }

    /**
     * Whether two vertices have the same current root. O(w log n) worst case. Throws `RangeError`
     * when either vertex is outside the fixed universe.
     */
    public connected(left: number, right: number): boolean {
        this.#checkVertices(left, right);
        return this.#findRoot(left) === this.#findRoot(right);
    }

    /**
     * The number of vertices in a vertex's connected component. O(w log n) worst case.
     *
     * An isolated vertex is still a singleton root and reports 1. The answer is the union-by-size
     * count cached at the component's current root, so the walk reads the structure without
     * compressing it; this forest deliberately has no path compression. Throws `RangeError` when
     * the vertex is outside the fixed universe.
     */
    public componentSize(vertex: number): number {
        this.#checkVertex(vertex);
        return this.#sizeOf(this.#findRoot(vertex));
    }

    /**
     * A child history version containing an undirected connection between two vertices.
     *
     * The returned version is always distinct, and is never the receiver. When the endpoints were
     * already connected — a self link included — its connectivity cells are shared unchanged, which
     * `sharesConnectivityRootWith` can confirm; otherwise exactly one union-by-size root edge is
     * added and labelled with the child version. On a size tie the first endpoint's root stays the
     * representative.
     *
     * O(w log n) time and O(w) new CHAMP nodes for a successful union; O(w log n) time and O(1)
     * space when redundant. The receiver is left untouched, including on failure: `RangeError` for
     * an endpoint outside the fixed universe is raised before any version is published.
     */
    public link(left: number, right: number): PersistentAncestralConnectionForest {
        this.#checkVertices(left, right);
        let leftRoot = this.#findRoot(left);
        let rightRoot = this.#findRoot(right);
        const childVersion = publishChildVersion(this.#version);

        if (leftRoot === rightRoot) {
            return new PersistentAncestralConnectionForest(
                this.#vertexCount, this.#componentCount, this.#cells, childVersion);
        }

        let leftSize = this.#sizeOf(leftRoot);
        let rightSize = this.#sizeOf(rightRoot);
        if (leftSize < rightSize) {
            [leftRoot, rightRoot] = [rightRoot, leftRoot];
            [leftSize, rightSize] = [rightSize, leftSize];
        }

        // On a size tie the first endpoint's root remains the representative, making the result
        // deterministic without changing the logarithmic-height proof. The combined size cannot
        // overflow: the two components are disjoint, so their sum is at most the vertex universe.
        const cells = this.#cells
            .put(leftRoot, rootCell(leftRoot, leftSize + rightSize))
            .put(rightRoot, childCell(leftRoot, childVersion));
        return new PersistentAncestralConnectionForest(
            this.#vertexCount, this.#componentCount - 1, cells, childVersion);
    }

    /**
     * The earliest version on this version's root-to-current history in which the pair was
     * connected, or `undefined` when they are disconnected in this version.
     *
     * A vertex is connected to itself at the history root. The answer is computed from the two
     * current parent paths as the latest union-edge version strictly below their forest lowest
     * common ancestor; the version history is never searched. O(w log n) worst-case time and
     * O(log n) temporary path space, independent of history depth.
     *
     * The result is never a version on another branch, and a later merge of the pair's component
     * into a bigger one never replaces the pair's earlier witness. The two failure modes stay
     * distinguishable: a vertex outside the fixed universe throws `RangeError`, while a pair that is
     * merely disconnected in this version returns `undefined`.
     */
    public firstConnected(left: number, right: number): AncestralConnectionVersion | undefined {
        this.#checkVertices(left, right);
        if (left === right) return this.#version.root;

        const leftPath = this.#buildPath(left);
        const rightPath = this.#buildPath(right);
        if (leftPath[leftPath.length - 1]?.vertex !== rightPath[rightPath.length - 1]?.vertex) return undefined;

        // Delete the common rootward suffix. The remaining steps are precisely the two paths
        // strictly below the forest LCA, and every one of those steps owns a successful-union tag.
        let leftLast = leftPath.length - 1;
        let rightLast = rightPath.length - 1;
        while (leftLast >= 0 && rightLast >= 0 && leftPath[leftLast]?.vertex === rightPath[rightLast]?.vertex) {
            leftLast--;
            rightLast--;
        }

        let latest: AncestralConnectionVersion | undefined;
        for (let index = 0; index <= leftLast; index++) latest = later(latest, leftPath[index]?.joinedAt);
        for (let index = 0; index <= rightLast; index++) latest = later(latest, rightPath[index]?.joinedAt);

        // Distinct connected vertices have at least one edge strictly below their LCA.
        if (latest === undefined) throw new Error("A connected pair has no union-edge witness.");
        return latest;
    }

    /**
     * Whether two versions retain the exact same sparse connectivity index, so neither can observe a
     * union recorded in the other. A representation test, not an equality test. O(1).
     */
    public sharesConnectivityRootWith(other: PersistentAncestralConnectionForest): boolean {
        return this.#cells.sharesRootWith(other.#cells);
    }

    /**
     * Walk the whole representation and check its invariants, throwing `Error` on the first
     * violation, then return its shape measurements. A defensive audit, not part of normal use.
     *
     * Checks the cached counts, the history-token parent chain and root, sparse-cell canonicality,
     * union-tag ancestry and order, the union-by-size height ceiling of `floor(log2 n)`, the cached
     * component count, and the cached root sizes. O(H + m w log n), where H is the history depth and
     * m is the explicit-cell count.
     */
    public validateStructure(): AncestralConnectionForestStatistics {
        if (this.#vertexCount < 0 || this.#componentCount < 0 || this.#componentCount > this.#vertexCount) {
            throw new Error("Connection-forest counts are invalid.");
        }

        const historyRoot = this.#version.root;
        const ancestors = new Set<AncestralConnectionVersion>();
        let expectedDepth = this.#version.depth;
        let version: AncestralConnectionVersion | undefined = this.#version;
        while (version !== undefined) {
            if (version.depth !== expectedDepth || version.root !== historyRoot) {
                throw new Error("The history-token parent chain is inconsistent.");
            }
            expectedDepth--;
            ancestors.add(version);
            version = version.parent;
        }
        if (expectedDepth !== -1 || !ancestors.has(historyRoot)
            || historyRoot.parent !== undefined || historyRoot.depth !== 0) {
            throw new Error("The history root is inconsistent.");
        }

        for (const entry of this.#cells) {
            const cell = entry.value;
            if (!this.#isVertex(entry.key) || !this.#isVertex(cell.parent)) {
                throw new Error("A sparse parent cell addresses a vertex outside the universe.");
            }
            if (cell.parent === entry.key) {
                if (cell.size <= 1 || cell.joinedAt !== undefined) {
                    throw new Error("A stored root cell is not a canonical non-singleton root.");
                }
            } else if (cell.size !== 0 || cell.joinedAt === undefined || !ancestors.has(cell.joinedAt)
                || !this.#cells.containsKey(cell.parent)) {
                throw new Error("A non-root cell has an invalid size or union-version tag.");
            }
        }

        const componentSizes = new Map<number, number>();
        let maximumParentPathLength = 0;
        const maximumHeight = this.#vertexCount <= 1 ? 0 : 31 - Math.clz32(this.#vertexCount);
        for (const entry of this.#cells) {
            let current = entry.key;
            let height = 0;
            let previousDepth = -1;
            for (;;) {
                const cell = this.#cell(current);
                if (cell.parent === current) break;
                const joinedAt = cell.joinedAt;
                if (joinedAt === undefined || joinedAt.depth <= previousDepth || !ancestors.has(joinedAt)) {
                    throw new Error("Union-edge versions do not strictly increase toward the root.");
                }
                previousDepth = joinedAt.depth;
                current = cell.parent;
                if (++height > maximumHeight) {
                    throw new Error("A parent path exceeds the union-by-size height bound.");
                }
            }
            maximumParentPathLength = Math.max(maximumParentPathLength, height);
            componentSizes.set(current, (componentSizes.get(current) ?? 0) + 1);
        }

        const storedCellCount = this.#cells.size;
        const absentSingletons = this.#vertexCount - storedCellCount;
        if (absentSingletons < 0 || absentSingletons + componentSizes.size !== this.#componentCount) {
            throw new Error("The cached component count disagrees with the parent forest.");
        }
        for (const [root, size] of componentSizes) {
            if (this.#sizeOf(root) !== size) throw new Error("A root's cached union size is incorrect.");
        }

        return {
            vertexCount: this.#vertexCount,
            componentCount: this.#componentCount,
            storedCellCount,
            maximumParentPathLength,
            historyDepth: this.#version.depth,
        };
    }

    #findRoot(vertex: number): number {
        let current = vertex;
        for (;;) {
            const parent = this.#parentOf(current);
            if (parent === current) return current;
            current = parent;
        }
    }

    #buildPath(vertex: number): PathStep[] {
        const path: PathStep[] = [];
        let current = vertex;
        for (;;) {
            const cell = this.#cells.get(current);
            const parent = cell === undefined ? current : cell.parent;
            path.push({ vertex: current, joinedAt: cell?.joinedAt });
            if (parent === current) return path;
            current = parent;
        }
    }

    /** The cell of a vertex; an absent cell is canonically its own singleton root. */
    #cell(vertex: number): Cell { return this.#cells.get(vertex) ?? rootCell(vertex, 1); }

    #parentOf(vertex: number): number { return this.#cells.get(vertex)?.parent ?? vertex; }

    #sizeOf(vertex: number): number { return this.#cells.get(vertex)?.size ?? 1; }

    #isVertex(vertex: number): boolean {
        return Number.isInteger(vertex) && vertex >= 0 && vertex < this.#vertexCount;
    }

    #checkVertex(vertex: number): void {
        if (!this.#isVertex(vertex)) {
            throw new RangeError(
                `Vertex ${vertex} must lie within the forest's fixed universe of ${this.#vertexCount} vertices.`);
        }
    }

    #checkVertices(left: number, right: number): void {
        this.#checkVertex(left);
        this.#checkVertex(right);
    }
}

File: src/OCaml/lib/hamt/persistent_ancestral_connection_forest.ml
(** Implementation of the fully branching persistent insertion-only connectivity forest.

    Parent cells live sparsely in a CHAMP map keyed by vertex, an absent cell meaning "singleton
    root of size 1". Union by size without path compression keeps every parent path within
    [floor (log2 vertex_count)] edges, and the one root-parent edge each successful link adds
    carries the child version, which is what turns two current parent paths into a first-connection
    witness without ever searching the version history.

    Every chain walk here is a loop or a tail call. Parent paths are logarithmic and could afford
    recursion, but a version chain cannot: histories run hundreds of thousands of links deep, and a
    non-tail walk over one would exhaust the stack. *)

type version = {
  version_parent : version option;
  version_depth : int;
  mutable root_link : version option;
      (* The history root of this version's tree, [None] only for a history root between its
         allocation and the assignment that ties its self-reference, which no caller can observe.
         The field is mutable for exactly that knot, and never assigned again. It is also what keeps
         a history-root record from being a compile-time constant: the native compiler lifts those
         into shared static data, which would make two independently created forests share one
         history root and defeat identity comparison. *)
}

(* One sparse parent record. [cell_size] is meaningful only for a root, that is when [cell_parent]
   is the vertex itself, and [joined_at] is present only on a non-root edge. *)
type cell = { cell_parent : int; cell_size : int; joined_at : version option }

type t = {
  universe : int; (* the fixed vertex-universe size *)
  components : int; (* the cached connected-component count *)
  cells : (int, cell) Persistent_hamt.map;
  current_version : version;
}

type statistics = {
  vertex_count : int;
  component_count : int;
  stored_cell_count : int;
  maximum_parent_path_length : int;
  history_depth : int;
}

(* Hashing a vertex to itself is what earns the unconditional cell-access bound: vertices are
   validated into [0 .. universe - 1] before any cell is touched, so the trie's [land max_int] mask
   is the identity and the hash is injective over the whole key universe. No collision bucket can
   then hold two distinct keys, and two distinct vertices diverge within the 62 significant hash
   bits, that is within 13 levels of 32-way nodes. *)
let vertex_policy : int Common.Hash_policy.t =
  Common.Hash_policy.create ~hash:(fun vertex -> vertex) ~equal:Int.equal

(* Cells are compared field by field, with version tags compared by identity. The map's default
   structural equality would instead walk a version's whole parent chain, and would not terminate at
   all on the history root's self-reference. *)
let cell_equal left right =
  left.cell_parent = right.cell_parent
  && left.cell_size = right.cell_size
  &&
  match (left.joined_at, right.joined_at) with
  | None, None -> true
  | Some left_tag, Some right_tag -> left_tag == right_tag
  | None, Some _ | Some _, None -> false

let version_root token = match token.root_link with Some root -> root | None -> token

let fresh_history_root () =
  let root = { version_parent = None; version_depth = 0; root_link = None } in
  root.root_link <- Some root;
  root

let child_version parent =
  if parent.version_depth = max_int then
    Error "the connection-forest history depth would overflow"
  else
    Ok
      {
        version_parent = Some parent;
        version_depth = parent.version_depth + 1;
        root_link = Some (version_root parent);
      }

let create vertex_count =
  if vertex_count < 0 then Error "vertex count cannot be negative"
  else
    Ok
      {
        universe = vertex_count;
        components = vertex_count;
        cells = Persistent_hamt.empty ~value_equal:cell_equal vertex_policy;
        current_version = fresh_history_root ();
      }

let vertex_count forest = forest.universe
let component_count forest = forest.components
let version forest = forest.current_version
let version_depth token = token.version_depth
let version_parent token = token.version_parent
let same_version left right = left == right
let in_universe forest vertex = vertex >= 0 && vertex < forest.universe

let check_vertex forest vertex =
  if in_universe forest vertex then Ok ()
  else Error "vertex must lie within the forest's fixed universe"

let check_vertices forest left right =
  match check_vertex forest left with
  | Error _ as error -> error
  | Ok () -> check_vertex forest right

let cell_at forest vertex =
  match Persistent_hamt.find_opt vertex forest.cells with
  | Some cell -> cell
  | None -> { cell_parent = vertex; cell_size = 1; joined_at = None }

(* An absent cell is canonically its own singleton root. *)
let parent_of forest vertex =
  match Persistent_hamt.find_opt vertex forest.cells with
  | Some cell -> cell.cell_parent
  | None -> vertex

let size_of forest vertex =
  match Persistent_hamt.find_opt vertex forest.cells with
  | Some cell -> cell.cell_size
  | None -> 1

let rec find_root forest vertex =
  let parent = parent_of forest vertex in
  if parent = vertex then vertex else find_root forest parent

let find vertex forest =
  match check_vertex forest vertex with
  | Error _ as error -> error
  | Ok () -> Ok (find_root forest vertex)

let connected left right forest =
  match check_vertices forest left right with
  | Error _ as error -> error
  | Ok () -> Ok (find_root forest left = find_root forest right)

let component_size vertex forest =
  match check_vertex forest vertex with
  | Error _ as error -> error
  | Ok () -> Ok (size_of forest (find_root forest vertex))

let link left right forest =
  match check_vertices forest left right with
  | Error _ as error -> error
  | Ok () -> (
      match child_version forest.current_version with
      | Error message -> Error message
      | Ok child ->
          let left_root = find_root forest left in
          let right_root = find_root forest right in
          if left_root = right_root then Ok { forest with current_version = child }
          else
            let left_size = size_of forest left_root in
            let right_size = size_of forest right_root in
            (* On a size tie the first endpoint's root remains the representative, making the
               result deterministic without changing the logarithmic-height proof. *)
            let keeper, absorbed =
              if left_size < right_size then (right_root, left_root) else (left_root, right_root)
            in
            let cells =
              forest.cells
              |> Persistent_hamt.set keeper
                   { cell_parent = keeper; cell_size = left_size + right_size; joined_at = None }
              |> Persistent_hamt.set absorbed
                   { cell_parent = keeper; cell_size = 0; joined_at = Some child }
            in
            Ok { forest with components = forest.components - 1; cells; current_version = child })

(* The parent path of a vertex, as [(vertex, joined_at)] steps ordered from the current root down to
   the vertex itself. Never empty. *)
let build_path forest vertex =
  let rec walk current accumulated =
    let cell = cell_at forest current in
    let accumulated = (current, cell.joined_at) :: accumulated in
    if cell.cell_parent = current then accumulated else walk cell.cell_parent accumulated
  in
  walk vertex []

(* Keeps the deeper of two candidate witnesses, and the left one on equal depth. *)
let later best candidate =
  match (best, candidate) with
  | _, None -> best
  | None, Some _ -> candidate
  | Some current, Some proposed ->
      if proposed.version_depth > current.version_depth then candidate else best

let rec drop_shared_prefix left right =
  match (left, right) with
  | (left_vertex, _) :: left_rest, (right_vertex, _) :: right_rest when left_vertex = right_vertex
    ->
      drop_shared_prefix left_rest right_rest
  | _ -> (left, right)

let first_connected left right forest =
  match check_vertices forest left right with
  | Error _ as error -> error
  | Ok () ->
      if left = right then Ok (Some (version_root forest.current_version))
      else
        let left_path = build_path forest left in
        let right_path = build_path forest right in
        let same_component =
          match (left_path, right_path) with
          | (left_root, _) :: _, (right_root, _) :: _ -> left_root = right_root
          | [], _ | _, [] -> false
        in
        if not same_component then Ok None
        else
          (* Delete the common rootward prefix. The remaining steps are precisely the two paths
             strictly below the forest LCA, and every one of those steps owns a union tag. *)
          let left_rest, right_rest = drop_shared_prefix left_path right_path in
          let latest = List.fold_left (fun best (_, tag) -> later best tag) None left_rest in
          let latest = List.fold_left (fun best (_, tag) -> later best tag) latest right_rest in
          (* Distinct connected vertices have at least one edge strictly below their LCA. *)
          if Option.is_none latest then Error "a connected pair has no union-edge witness"
          else Ok latest

let shares_connectivity_root_with left right = left.cells == right.cells
let rec floor_log2 value = if value <= 1 then 0 else 1 + floor_log2 (value / 2)

let validate forest =
  if forest.universe < 0 || forest.components < 0 || forest.components > forest.universe then
    Error "connection-forest counts are invalid"
  else
    let history_depth = forest.current_version.version_depth in
    let history_root = version_root forest.current_version in
    if history_depth < 0 then Error "the history-token parent chain is inconsistent"
    else begin
      (* One lineage slot per depth: the root-to-current chain holds exactly one version at each
         depth, so this indexes ancestry in constant time without hashing a token. *)
      let lineage = Array.make (history_depth + 1) forest.current_version in
      let is_ancestor token =
        let depth = token.version_depth in
        depth >= 0 && depth <= history_depth && lineage.(depth) == token
      in
      let rec walk_lineage token expected =
        if
          expected < 0 || token.version_depth <> expected
          || not (version_root token == history_root)
        then Error "the history-token parent chain is inconsistent"
        else begin
          lineage.(expected) <- token;
          match token.version_parent with
          | Some parent -> walk_lineage parent (expected - 1)
          | None -> Ok expected
        end
      in
      let stored = Persistent_hamt.to_list forest.cells in
      let rec check_cells = function
        | [] -> Ok ()
        | (vertex, cell) :: rest ->
            if not (in_universe forest vertex && in_universe forest cell.cell_parent) then
              Error "a sparse parent cell addresses a vertex outside the universe"
            else if cell.cell_parent = vertex then
              if cell.cell_size <= 1 || Option.is_some cell.joined_at then
                Error "a stored root cell is not a canonical non-singleton root"
              else check_cells rest
            else if
              cell.cell_size <> 0
              || (match cell.joined_at with Some tag -> not (is_ancestor tag) | None -> true)
              || not (Persistent_hamt.mem cell.cell_parent forest.cells)
            then Error "a non-root cell has an invalid size or union-version tag"
            else check_cells rest
      in
      let maximum_height = if forest.universe <= 1 then 0 else floor_log2 forest.universe in
      let rec climb current height previous_depth =
        let cell = cell_at forest current in
        if cell.cell_parent = current then Ok (current, height)
        else
          match cell.joined_at with
          | None -> Error "union-edge versions do not strictly increase toward the root"
          | Some tag ->
              if tag.version_depth <= previous_depth || not (is_ancestor tag) then
                Error "union-edge versions do not strictly increase toward the root"
              else if height + 1 > maximum_height then
                Error "a parent path exceeds the union-by-size height bound"
              else climb cell.cell_parent (height + 1) tag.version_depth
      in
      let component_sizes = Hashtbl.create 16 in
      let longest_path = ref 0 in
      let rec walk_paths = function
        | [] -> Ok ()
        | (vertex, _) :: rest -> (
            match climb vertex 0 (-1) with
            | Error message -> Error message
            | Ok (root, height) ->
                if height > !longest_path then longest_path := height;
                let counted = Option.value (Hashtbl.find_opt component_sizes root) ~default:0 in
                Hashtbl.replace component_sizes root (counted + 1);
                walk_paths rest)
      in
      let rec check_sizes = function
        | [] -> Ok ()
        | (root, size) :: rest ->
            if size_of forest root <> size then Error "a root's cached union size is incorrect"
            else check_sizes rest
      in
      match walk_lineage forest.current_version history_depth with
      | Error message -> Error message
      | Ok last_depth ->
          if
            last_depth <> 0
            || not (is_ancestor history_root)
            || Option.is_some history_root.version_parent
            || history_root.version_depth <> 0
          then Error "the history root is inconsistent"
          else (
            match check_cells stored with
            | Error message -> Error message
            | Ok () -> (
                match walk_paths stored with
                | Error message -> Error message
                | Ok () ->
                    let stored_cell_count = Persistent_hamt.count forest.cells in
                    let absent_singletons = forest.universe - stored_cell_count in
                    if
                      absent_singletons < 0
                      || absent_singletons + Hashtbl.length component_sizes <> forest.components
                    then Error "the cached component count disagrees with the parent forest"
                    else
                      let roots =
                        Hashtbl.fold (fun root size found -> (root, size) :: found)
                          component_sizes []
                      in
                      (match check_sizes roots with
                      | Error message -> Error message
                      | Ok () ->
                          Ok
                            {
                              vertex_count = forest.universe;
                              component_count = forest.components;
                              stored_cell_count;
                              maximum_parent_path_length = !longest_path;
                              history_depth;
                            })))
    end
